apps/antalmanac/src/actions/ActionTypesStore.ts

```diff
@@ -0,0 +1,151 @@
+import { EventEmitter } from 'events';
+import { RepeatingCustomEvent, ScheduleCourse } from '@packages/antalmanac-types';
+import { saveSchedule } from '$actions/AppStoreActions';
+import AppStore from '$stores/AppStore';
+
+export interface BaseAction {
+    type: string;
+}
+
+export interface UndoAction extends BaseAction {
+    type: 'undoAction';
+}
+
+export interface AddCourseAction extends BaseAction {
+    type: 'addCourse';
+    course: ScheduleCourse;
+    scheduleIndex: number;
+}
+
+export interface DeleteCourseAction extends BaseAction {
+    type: 'deleteCourse';
+    sectionCode: string;
+    term: string;
+}
+
+export interface AddCustomEventAction extends BaseAction {
+    type: 'addCustomEvent';
+    customEvent: RepeatingCustomEvent;
+    scheduleIndices: number[];
+}
+
+export interface DeleteCustomEventAction extends BaseAction {
+    type: 'deleteCustomEvent';
+    customEventId: number;
+}
+
+export interface EditCustomEventAction extends BaseAction {
+    type: 'editCustomEvent';
+    editedCustomEvent: RepeatingCustomEvent;
+    newScheduleIndices: number[];
+}
+
+export interface ChangeCustomEventColorAction extends BaseAction {
+    type: 'changeCustomEventColor';
+    customEventId: number;
+    newColor: string;
+}
+
+export interface ClearScheduleAction extends BaseAction {
+    type: 'clearSchedule';
+}
+
+export interface CopyScheduleAction extends BaseAction {
+    type: 'copySchedule';
+    to: number;
+}
+
+export interface ChangeCourseColorAction extends BaseAction {
+    type: 'changeCourseColor';
+    sectionCode: string;
+    term: string;
+    newColor: string;
+}
+
+export type ActionType =
+    | AddCourseAction
+    | DeleteCourseAction
+    | AddCustomEventAction
+    | DeleteCustomEventAction
+    | EditCustomEventAction
+    | ChangeCustomEventColorAction
+    | ClearScheduleAction
+    | CopyScheduleAction
+    | ChangeCourseColorAction
+    | UndoAction;
+
+class ActionTypesStore extends EventEmitter {
+    constructor() {
+        super();
+    }
+
+    async autoSaveSchedule(action: ActionType) {
+        const autoSave = typeof Storage !== 'undefined' && window.localStorage.getItem('autoSave') == 'true';
+        if (autoSave) {
+            const savedUserID = window.localStorage.getItem('userID');
+
+            if (savedUserID) {
+                this.emit('autoSaveStart');
+                await saveSchedule(savedUserID, true, true);
+                AppStore.unsavedChanges = false;
+                this.emit('autoSaveEnd');
+            }
+        } else {
+            const unsavedActionsString = window.localStorage.getItem('unsavedActions');
+            if (unsavedActionsString == null) {
+                const unsavedActions = [action];
+                localStorage.setItem('unsavedActions', JSON.stringify(unsavedActions));
+            } else {
+                const unsavedActions = JSON.parse(unsavedActionsString);
+                unsavedActions.push(action);
```

Comment:
You can also use a type-safe parsing function here, like I mentioned in a different comment. Also, could you add an upper limit to the length of actions? It can be arbitrary for now, but I think it should conceptually exist. For example, a max of 1000 unsaved actions can be set.

```ts
// At the top of the module.
const MAX_UNSAVED_ACTIONS = 1_000

const unsavedActions = parseUnsavedActionsString(unsavedActionsString)
if (unsavedActions.length > MAX_UNSAVED_ACTIONS) {
  // Do something, e.g. slice of the earliest 100?
}

// If max was reached, and resolved by removing some items, 
// then you could proceed as normal
unsavedActions.push(action)
```